[vcpkg] Fix "[commands-build] build smoke test"

ras0219 opened the pull request:
1. Do not exit with a success code in Build::perform, because this causes failures in other tests to be ignored.
2. Use temp directory to avoid interference with the current set of {installed, buildtrees, packages}
3. Explicitly disable binary caching

- Which triplets are supported/not supported? Have you updated the CI baseline?

N/A

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

Yes

ras0219-msft committed:
[vcpkg] Fix "[commands-build] build smoke test"

1. Do not exit with a success code in Build::perform, because this causes failures in other tests to be ignored.
2. Use temp directory to avoid interference with the current set of {installed, buildtrees, packages}
3. Explicitly disable binary caching

strega-nil commented on toolsrc/src/vcpkg/build.cpp:
```diff
@@ -143,7 +143,7 @@ namespace vcpkg::Build
             Checks::exit_fail(VCPKG_LINE_INFO);
```
Could you also change these to `return 1;`?

ras0219-msft committed:
[vcpkg] Disable binarycaching in "build smoke test", "[commands-build]"

Return 1 for the last failure case in Build::perform() so testing may continue.

strega-nil commented on toolsrc/src/vcpkg-test/commands.build.cpp:
```diff
@@ -6,15 +6,27 @@
     VcpkgCmdArguments args = VcpkgCmdArguments::create_from_arg_sequence(std::begin(args_raw), std::end(args_raw));
+    args.binary_caching = false;
+    args.buildtrees_root_dir =
+        std::make_unique<std::string>((Test::base_temporary_directory() / fs::u8path("buildtrees")).u8string());
```
Could we use `tmp/<some random string>` here? Maybe `Test::base_temporary_directory() / get_uuid() / fs::u8path("buildtrees")`?

ras0219 replied:
The reason I went with a non-random string is because the symlink tests are also using a fixed seed (from what I can tell). This would indicate that our tests are non-concurrent, so this is simple.

strega-nil reviewed (changes requested):
Just the one change, then approved!
